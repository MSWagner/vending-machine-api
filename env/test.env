NODE_ENV=test

PGHOST=localhost
PGPORT=5431
PGUSER=root
PGPASSWORD=toor
PGDATABASE=vending-machine-api-test-db