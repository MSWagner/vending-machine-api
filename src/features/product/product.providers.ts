import { Connection, Repository } from "typeorm";
import { Product } from "../../entities/Product.entity";

import CONFIG from "../../config";

export const productProviders = [
    {
        provide: CONFIG.database.defaultProductRepoName,
        useFactory: (connection: Connection): Repository<Product> => connection.getRepository(Product),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
