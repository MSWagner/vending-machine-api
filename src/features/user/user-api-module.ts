import { Module } from "@nestjs/common";
import { UserController } from "./controller/user.controller";
import { AuthModule } from "../auth/auth.module";
import { UserModule } from "./user.module";

@Module({
    imports: [AuthModule, UserModule],
    controllers: [UserController]
})
export class UserApiModule {}
