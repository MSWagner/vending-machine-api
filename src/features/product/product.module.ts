import { Module } from "@nestjs/common";
import { ProductService } from "./services/product.service";
import { productProviders } from "./product.providers";

import { DatabaseModule } from "../../shared/database/database.module";
import { UserModule } from "../user/user.module";
import { AuthModule } from "../auth/auth.module";
import { ProductController } from "./controller/product.controller";

@Module({
    imports: [DatabaseModule, UserModule, AuthModule],
    providers: [...productProviders, ProductService],
    controllers: [ProductController],
    exports: [DatabaseModule, ProductService, ...productProviders]
})
export class ProductModule {}
