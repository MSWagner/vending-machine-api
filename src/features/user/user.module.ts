import { Module } from "@nestjs/common";
import { UserService } from "./services/user.service";
import { userProviders } from "./user.providers";

import { DatabaseModule } from "../../shared/database/database.module";

@Module({
    imports: [DatabaseModule],
    providers: [...userProviders, UserService],
    exports: [DatabaseModule, UserService, ...userProviders]
})
export class UserModule {}
