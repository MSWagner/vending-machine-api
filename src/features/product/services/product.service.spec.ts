import { Test, TestingModule } from "@nestjs/testing";

import { TestModule } from "../../../shared/test/test.module";
import { TestService } from "../../../shared/test/test.service";

import { ProductService } from "./product.service";
import { productProviders } from "../product.providers";
import { Product } from "../../../entities/Product.entity";

import { User } from "../../../entities/User.entity";
import { UserModule } from "../../../features/user/user.module";

import * as fixtures from "../../../shared/test/fixtures";

describe("ProductService", () => {
    let service: ProductService;
    let testService: TestService;

    beforeEach(async () => {
        const module: TestingModule = await Test.createTestingModule({
            imports: [TestModule, UserModule],
            providers: [...productProviders, ProductService]
        }).compile();

        service = module.get<ProductService>(ProductService);
        testService = module.get<TestService>(TestService);

        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    it("should find product1", async () => {
        const product1 = await service.findOneProduct(fixtures.product1User3Seller.uid);

        expect(product1.uid).toEqual(fixtures.product1User3Seller.uid);
    });

    it("should not find some product with unknown product uuid", async () => {
        const user = await service.findOneProduct("ad9ee500-c03e-4d4e-9ef2-3669ab17b023");

        expect(user).toBeUndefined();
    });

    it("should create a new product", async () => {
        const seller = await User.findOne(fixtures.user3.uid);

        const productCountBefore = await Product.count({ where: { productName: "testProduct1" } });
        expect(productCountBefore).toEqual(0);

        const newProduct = await service.createProduct(seller, "testProduct1", 10, 100);
        expect(newProduct.productName).toEqual("testProduct1");
        expect(newProduct.cost).toEqual(10);
        expect(newProduct.amountAvailable).toEqual(100);
        expect(newProduct.seller.username).toEqual(seller.username);

        const productCountAfter = await Product.count({ where: { productName: "testProduct1" } });
        expect(productCountAfter).toEqual(1);
    });

    it("should update a product", async () => {
        const product1 = await Product.findOne(fixtures.product1User3Seller.uid);
        const seller = await User.findOne(fixtures.user3.uid);

        const productCountBefore = await Product.count({ where: { productName: "testProduct1" } });
        expect(productCountBefore).toEqual(0);

        const newProduct = await service.updateUser(product1, seller, "testProduct1Updated", 20, 200);
        expect(newProduct.uid).toEqual(product1.uid);
        expect(newProduct.productName).toEqual("testProduct1Updated");
        expect(newProduct.cost).toEqual(20);
        expect(newProduct.amountAvailable).toEqual(200);
        expect(newProduct.seller.username).toEqual(seller.username);

        const productCountAfter = await Product.count({ where: { productName: "testProduct1Updated" } });
        expect(productCountAfter).toEqual(1);
    });

    it("should delete product1", async () => {
        const product1 = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(product1).not.toBeUndefined();

        await service.deleteUser(product1);

        const removedProduct = await Product.findOne(fixtures.product1User3Seller.uid);
        expect(removedProduct).toBeUndefined();
    });
});
