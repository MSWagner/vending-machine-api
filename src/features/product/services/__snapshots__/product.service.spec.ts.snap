// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductService should find all products: AllProducts 1`] = `
[
  Product {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  Product {
    "amountAvailable": 1,
    "cost": 10,
    "createdAt": "createdAt",
    "productName": "product2",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
  Product {
    "amountAvailable": 10,
    "cost": 100,
    "createdAt": "createdAt",
    "productName": "product1Seller4",
    "uid": "uid",
    "updatedAt": "updatedAt",
  },
]
`;
